File: languages/go/bits.go
// bits.go

// Print all bit strings of length n. It should work for any integer n >= 0,
// although since there are an exponential number of bit strings it's okay if
// your program takes a long time to calculate large values of n.

package main

import "fmt"

func main() {
    n := 4
    fmt.Printf("%v-bit strings:\n", n)
    bits := nbits(n)
    for i, b := range bits {
        fmt.Printf("%v. %v\n", i+1, b)
    }
}

// Returns a slice of all bit-strings of length n.
func nbits(n int) []string {
    if n < 0 {
        return []string{}
    } else if n == 1 {
        return []string{"0", "1"}
    } else {
        n1bits := nbits(n - 1)
        zero := append([]string{}, n1bits...)
        one := append([]string{}, n1bits...)

        for i := range n1bits {
            zero[i] = "0" + zero[i]
            one[i] = "1" + one[i]
        }
        return append(zero, one...)
    }
}

File: languages/ruby/bits.rb
# bits.rb

#
# Print all bit strings of length n, for any integer n >= 0. There are an
# exponential number of bit strings and so it's okay if your program takes a
# long time to calculate large values of n.
#
# The general strategy used in nbits is to calculate all n-bit strings is to
# first recursively make two copies of all (n-1)-bit strings. To one copy '0'
# is appended the start of each bit strings, and to the other '1' is appended.
# Then the two lists are appended together to get the final result.
#
def nbits(n)
	# non-recursive base cases
	return [] if n < 0
	return ['0','1'] if n == 1

	# recursive case
	n1bits = nbits(n-1)
	zero = n1bits.map {|s| '0' + s}
	one = n1bits.map {|s| '1' + s}
	return zero + one
end

#
# Example
#
results = nbits(5)
results.each_with_index do |b,i|
	puts "#{b} #{i}"
end
